File: StaticSharp/BasicComponents/AspectBlock/AspectBlock.cs
using StaticSharp.Gears;

namespace StaticSharp {
    public enum Fit { 
        Inside,
        Outside,
        Stretch
    }

    public interface JAspectBlock : JBlock {
        public double NativeAspect { get; set; }
        public Fit Fit { get; set; }
        public double GravityVertical { get; set; }
        public double GravityHorizontal { get; set; }
    }

    [ConstructorJs]
    public abstract partial class AspectBlock : Block {
        /*protected static void SetNativeSize(Scopes.Group script, string elementVariableName, double width, double height) {
            script.Add($"{elementVariableName}.NativeWidth = {width.ToStringInvariant()}");
            script.Add($"{elementVariableName}.NativeHeight = {height.ToStringInvariant()}");
        }*/
    }
}

File: StaticSharp/BasicComponents/AspectBlock/AspectBlock.js




function AspectBlock(element) {
    Block(element)

    element.Reactive = {

        NativeWidth: undefined,
        NativeHeight: undefined,
        NativeAspect: e => e.NativeWidth / e.NativeHeight,

        //Aspect: e => e.NativeAspect,
        Width: () => Num.First(element.Height * element.NativeAspect, element.NativeWidth),
        Height: () => Num.First(element.Width / element.NativeAspect, element.NativeHeight),

        Fit: "Inside",
        GravityVertical: 0,
        GravityHorizontal: 0,
    }

    


}